### ISSUE TITLE
Profiling Suite Runtimes

### ISSUE THREAD

**Issue by onsi** [onsi/ginkgo#861]
@KevinJCross wrote on #711 

Hi, Im considering trying to add a feature where a file log of the time of each stage is aded. Currently we have quite a few issues with slowness in our meduim sized test suites. To try optimise this we would like to get some metrics as to which bits are taking the most time to focus our approach.

I.e.
CompileTimes
BeforeSuite
AfterSuite
BeforeEach's
AfterEach's
JustBefore
and Its
Practically anywhere where time is spent inside or outside the test suite.
I recognise that most/all of it would be in the tests but to pinpoint which justbefore each or BeforeEach step call is the most called and which ones run the would help gather the information we would need to effectively optimise our test suites.

I would thikning along the lines of a csv file output styles with something like:

Description, average call time (s), total time (s), number of calls 
"suite"."beforeSuite".BeforeEach, 10, 10, 1
"suite"."Description"."context"."context".BeforeEach, 0.200, 20, 200
There is the possiblity we could farm it out of the cpu performance logs I think this would be usefull to owners of larger suites that are taking too long and would like to cut the runtimes down, So including it as a feature might be a good idea.

**Comment by onsi** [comment]
hey @KevinJCross some thoughts on this.  The use case makes sense to me.  I can imagine a few ways of approaching this once 2.0 is out (I'd like to get 2.0 out first - it's just about ready at this point).

I'm a bit hesitant to introduce an entirely new type of report complete with its own format and would rather build on top of the emerging 2.0 reporting infrastructure.  For example, a new `--measure-node-runtimes` CLI flag could instruct Ginkgo to measure the runtime for each node and annotate each spec report with a ReportEntry for each node in the spec and how long the node took to run.  This data would then be available in any JSON-formatted reports and users could parse the file and perform whatever computations they want.  Alternatively, I could imagine making "Instrumentation" a more full-fledged feature of Ginkgo.  Users can register `NodeInstrument`s that are called before and after each node runs.  These instruments could then generate and add ReportEntries to the report that contain arbitrary information (e.g. runtime, memory usage) and we could ship with a `RuntimeNodeInstrument` out of the box for this use case.

Any thoughts?

**Comment by KevinJCross** [comment]
@onsi sure I totally agree with getting 2.0 out and this can just be a feature say in say 2.1. I also figure that 2.0 is now just going through a `stabilisation` process so the architecture should be stable enough to start working on something like this for 2.1.

Yeah IU think maybe a visitor or observer-ish pattern or such using the nodes could be quite generic and usefull in the future.

> Users can register NodeInstrument

yeah this sounds like the type of thing I was going to look into and adding a `RuntimeNodeInstrument` which gets switched on with the `--measure-node-runtimes` or  similar CLI feature would be pretty much what I need.

yeah a json or yaml report would probably be better since we can easily manipulate it via jq or yq on the cli to post process the information ad hoc, maybe I could include a separate summary tool that takes the json and outputs a table as given in the above.

Soo ... Im very unfamiliar with the internals and the reporting in Ginkgo atm ... where do I start looking to get up to speed on how the tests/nodes start running and the reporting integrations that are now in 2.0 ?

**Comment by KevinJCross** [comment]
@onsi also When would be a decent time to create a fork to start looking into this ?

**Comment by onsi** [comment]
hey @KevinJCross - ok sounds like the api design makes sense to you.  I'm close to pushing out a commit to the `ver2` branch that formally bumps the major version - which means all the imports across all the files in Ginkgo are going to change from `github.com/onsi/ginkgo` to `github.com/onsi/ginkgo/v2` so I definitely wouldn't fork until then.

As for internals/reporting etc. I can talk through what an implementation could look like.  I'm also happy to do the work myself but if you'd like to give it a go I'm happy to point you in the right direction.  Another distribution of labor could also be for me to implement the node instrumentation bits and then you could add a subcommand to the cli tool that can analyze the runtime-profile in the generated report to produce the table?  Just throwing it out there as an option.

**Comment by KevinJCross** [comment]
@onsi ok hmm had a little bash at it this evening to see how things fitted together.
https://github.com/KevinJCross/ginkgo/tree/runtime_reports
Its probably a lil all over the place.... but its more to find out how things work in ginkgo.

**Comment by KevinJCross** [comment]
@onsi it seems my poc is a bit like a middleware/decorator/pointCut rather than a Instrument.
I think this would a more flexible way of doing this where the node is available to you before the call and after and you can even not call the node ... i.e. maybe some filtering or skipping 🤷

**Comment by onsi** [comment]
hey @KevinJCross thanks for exploring options here.  I've thought about it a bit more and I find my opinion shifting.  Rather than build out a new instrumentation abstraction (and deal with the various edge cases it opens up) I want to go back to the original problem statement.  "How do I profile the runtime for my Ginkgo suites?" and I think the `pprof` toolchain provides a lot of goodness that we'll want to build on.

With 2.0 Ginkgo supports all the existing `go test` profiles and knows how to manage them when running in parallel (this is non-trivial as each parallel process generates its own profile and all the profiles need to be merged into one) and knows how to manage them when running multiple test suites (something that `go test` itself doesn't actually do).  The more I think about it the more the right approach seems to be to introduce a new Ginkgo-only type of profile with a flag (e.g. `ginkgo --runtime-profile=out.prof`) that can generate a `pprof` formatted file... except here the hierarchy encoded in the file won't be the call stack, it'll be the spec tree.

That will allow us to leverage all of the existing `go tool pprof` mechanisms for visualizing and understanding the data _and_ it will give us the same kinds of hierarchical analysis tools that users are used to.

So - rather than build out an entire new abstraction I think we can simply have a flag that causes `suite.go` to generate a profile.  I haven't played with the profile format too much but that would be the next step, I think.

Thoughts?

**Comment by KevinJCross** [comment]
ok hmmm ive not played with the pprof stuff much but wont just turning it on catch all the the test suite running as well as the code its testing at the same time. I think I stated originally that we could leverage the pprof flags but write some scripts around it to mine the information correctly.
There is a significant drawback around using the current pprof format is that its sampled and quick nodes might not appear in the pprof files.
Are you suggesting to create a pprof file from our own instrumentation or to use the sampling pprof itself ?

**Comment by onsi** [comment]
hey sorry for the delay.  I was offline for 🦃 -day.

Yes I'm suggesting Ginkgo creates its own proof file using its own instrumentation rather than sampling.  `pprof` is "just" a hierarchical protobuf format that associates stacks with measurements (i.e. numbers).  The measurement for Ginkgo would be runtime and he stacks would be the spec tree hierarchy.

**Comment by KevinJCross** [comment]
ok hmmm sounds interesting... Ill have a look into this too.
Sounds better than putting it in the report tbh.

**Comment by onsi** [comment]
Huh.  I think the latest work on timelines in Ginkgo 2.5 also happens to solve for this.  The JSON output format now includes SpecEvents which list the entry and exit times for each node.  Tooling could be built on top of that to extract precisely these metrics.

Happy to dig into it more if you're still interested @KevinJCross  - but seeing as how this is ~a year old I'm going to go ahead and close the issue.
